domain/schema/model/sql/0022-resource.sql

```diff
@@ -29,103 +29,85 @@ INSERT INTO resource_state VALUES
 
 CREATE TABLE resource (
     uuid TEXT NOT NULL PRIMARY KEY,
-    application_uuid TEXT NOT NULL,
+    charm_uuid TEXT NOT NULL,
     name TEXT NOT NULL,
     revision INT,
     origin_type_id INT NOT NULL,
     state_id INT NOT NULL,
-    size INT,
-    hash TEXT,
-    hash_type_id TEXT,
     created_at TIMESTAMP NOT NULL,
-    CONSTRAINT fk_resource_application_uuid
-    FOREIGN KEY (application_uuid)
-    REFERENCES application (uuid),
+    -- last_polled is when the repository was last polled for new resource
+    -- revisions. Only set if resource_state is 1.
+    last_polled TIMESTAMP,
+    CONSTRAINT fk_charm
+    FOREIGN KEY (charm_uuid)
+    REFERENCES charm (uuid),
     CONSTRAINT fk_resource_name
     FOREIGN KEY (name)
-    REFERENCES resource_meta (name),
+    REFERENCES charm_resource (name),
     CONSTRAINT fk_resource_origin_type_id
     FOREIGN KEY (origin_type_id)
     REFERENCES resource_origin_type (id),
     CONSTRAINT fk_resource_state_id
     FOREIGN KEY (state_id)
-    REFERENCES resource_state (id),
-    CONSTRAINT fk_resource_hash_type_id
-    FOREIGN KEY (hash_type_id)
-    REFERENCES object_store_metadata_hash_type (id)
+    REFERENCES resource_state (id)
 );
 
-CREATE UNIQUE INDEX idx_resource ON resource (application_uuid, name, state_id);
-
-CREATE TABLE resource_meta (
+-- Link table for applications and their resources.
+CREATE TABLE application_resource (
+    resource_uuid TEXT NOT NULL PRIMARY KEY,
     application_uuid TEXT NOT NULL,
-    name TEXT NOT NULL,
-    type_id INT NOT NULL,
-    path TEXT,
-    description TEXT,
-    CONSTRAINT fk_resource_type_id
-    FOREIGN KEY (type_id)
-    REFERENCES charm_resource_kind (id),
-    PRIMARY KEY (application_uuid, name)
+    CONSTRAINT fk_application_uuid
+    FOREIGN KEY (application_uuid)
+    REFERENCES application (uuid),
+    CONSTRAINT fk_resource_uuid
+    FOREIGN KEY (resource_uuid)
+    REFERENCES resource (uuid)
 );
 
-CREATE UNIQUE INDEX idx_resource_meta ON resource_meta (application_uuid, name);
-
-CREATE TABLE resource_supplied_by_type (
+CREATE TABLE resource_retrieved_by_type (
     id INT PRIMARY KEY,
     name TEXT NOT NULL
 );
 
-CREATE UNIQUE INDEX idx_resource_supplied_by_type
-ON resource_supplied_by_type (name);
+CREATE UNIQUE INDEX idx_resource_retrieved_by_type
+ON resource_retrieved_by_type (name);
 
-INSERT INTO resource_supplied_by_type VALUES
+INSERT INTO resource_retrieved_by_type VALUES
 (0, 'user'),
 (1, 'unit'),
 (2, 'application');
 
-CREATE TABLE resource_supplied_by (
-    uuid TEXT NOT NULL PRIMARY KEY,
-    supplied_by_type_id INT NOT NULL,
-    -- Name is the entity who supplied the resource blob:
+CREATE TABLE resource_retrieved_by (
+    resource_uuid TEXT NOT NULL PRIMARY KEY,
+    retrieved_by_type_id INT NOT NULL,
+    -- Name is the entity who retrieved the resource blob:
     --   The name of the user who uploaded the resource.
```

Comment:
**suggestion:** this comment could be moved to above the table rather than just being on this field.

I still feel a bit funny about this table, I think the name is still slightly wrong. This table would be set when a resource is added, this can happen either when a user uploads the resource, or a charm requests the resource for the first time. Saying "retrived by" for the user uploading the charm doesn't quite make sense. Maybe it could be renamed to `resource_added_by`?

(I suggested the current name but looking at it now I'm not happy with it, sorry about that!)